test_data/pil/arith_improved.pil

```diff
@@ -29,8 +29,7 @@ namespace Arith(N);
     /// returns f(0) + f(1) + ... + f(length - 1)
     let sum = |length, f| fold(length, f, 0, |acc, e| acc + e);
 
-    // TODO the weird syntax is needed so that this is not classified as a constant column
-    let force_boolean = (|| |x| x * (1 - x) == 0)();
+    let force_boolean: expr -> constr = |x| x * (1 - x) == 0;
```

Comment:
actually, why is a `constr` not just an `expr`?

Reply:
because it compares to zero. If you want to keep everything an expression (implicitly assuming it is equated to zero in the end) you can also return an expression, but I think it is cleaner like this.

Reply:
so a `constr` is an expr that = 0?

Reply:
a constr is a `expr  = expr`.

Reply:
https://docs.powdr.org/pil/types.html#constraints